// node-vibrant/lib/image/node.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Bluebird = require("bluebird");
var base_1 = require("./base");
var Jimp = require("jimp");
var http = require("http");
var https = require("https");
var URL_REGEX = /^(\w+):\/\/.*/i;
var PROTOCOL_HANDLERS = {
    http: http, https: https
};
var NodeImage = /** @class */ (function (_super) {
    __extends(NodeImage, _super);
    function NodeImage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NodeImage.prototype._loadByProtocolHandler = function (handler, src) {
        return new Bluebird(function (resolve, reject) {
            handler.get(src, function (r) {
                var buf = new Buffer('');
                r.on('data', function (data) { return buf = Buffer.concat([buf, data]); });
                r.on('end', function () { return resolve(buf); });
                r.on('error', function (e) { return reject(e); });
            });
        });
    };
    NodeImage.prototype._loadFromPath = function (src) {
        var _this = this;
        var m = URL_REGEX.exec(src);
        if (m) {
            var protocol = m[1].toLocaleLowerCase();
            var handler = PROTOCOL_HANDLERS[protocol];
            if (!handler) {
                return Bluebird.reject(new Error("Unsupported protocol: " + protocol));
            }
            return this._loadByProtocolHandler(handler, src)
                .then(function (buf) { return _this._loadByJimp(buf); });
        }
        else {
            return this._loadByJimp(src);
        }
    };
    NodeImage.prototype._loadByJimp = function (src) {
        var _this = this;
        // NOTE: TypeScript doesn't support union type to overloads yet
        //       Use type assertion to bypass compiler error
        return new Bluebird(function (resolve, reject) {
            Jimp.read(src, function (err, result) {
                if (err)
                    return reject(err);
                _this._image = result;
                resolve(_this);
            });
        });
    };
    NodeImage.prototype.load = function (image) {
        if (typeof image === 'string') {
            return this._loadFromPath(image);
        }
        else if (image instanceof Buffer) {
            return this._loadByJimp(image);
        }
        else {
            return Bluebird.reject(new Error('Cannot load image from HTMLImageElement in node environment'));
        }
    };
    NodeImage.prototype.clear = function () {
    };
    NodeImage.prototype.update = function (imageData) {
    };
    NodeImage.prototype.getWidth = function () {
        return this._image.bitmap.width;
    };
    NodeImage.prototype.getHeight = function () {
        return this._image.bitmap.height;
    };
    NodeImage.prototype.resize = function (targetWidth, targetHeight, ratio) {
        this._image.resize(targetWidth, targetHeight);
    };
    NodeImage.prototype.getPixelCount = function () {
        var bitmap = this._image.bitmap;
        return bitmap.width * bitmap.height;
    };
    NodeImage.prototype.getImageData = function () {
        return this._image.bitmap;
    };
    NodeImage.prototype.remove = function () {
    };
    return NodeImage;
}(base_1.ImageBase));
exports.default = NodeImage;
